kythe/docs/schema-overview.txt

```diff
@@ -0,0 +1,421 @@
+
+Here we have two more parts of `kcd.go`, surrounding usage of the `Compile`
+function. The function is defined in `kcd.go`, and then referenced with a
+function call in `memdb.go`. The anchor of code for `Compile` in `kcd.go` has a
+`childof` edge pointing to the function semantic node for `Revisions`.
+
+=== [[FeatureOverrides]] Class/Interface Hierarchy & Overrides
+
+Inheritance relationships are captured by the `extends` and `satisfies` edges.
+Override relationships are expressed by the `overrides` edge. These
+relationships are illustrated by this example from Go (which uses the
+`satisfies` edge to express Go's implicit interface satisfaction). In this
+example, the `childof` edges capture the containment relationships between the
+methods and their types:
+
+[graphviz]
+------------------------------------------------------------------------
+digraph ClassH {
+  graph [fontname="helvetica"]
+  node  [fontname="helvetica";shape=box,fontsize=10]
+  edge  [fontname="helvetica",fontsize=10]
+
+  base [label="type Reader interface\n(kcd.go)"]
+  basemeth [label="function Revisions\n(kcd.go)"]
+  basemeth -> base [label="childof"]
+  derived [label="type DB struct\n(memdb.go)"]
+  derived -> base [label="satisfies"]
+  derivedmeth [label="function Revisions\n(memdb.go)"]
+  derivedmeth -> derived [label="childof"]
+  derivedmeth -> basemeth [label="overrides"]
+}
+------------------------------------------------------------------------
+
+`kcd.go`:
+
+[source,Go]
+----
+type Reader interface {
+  Revisions(_ context.Context, filter *RevisionsFilter, f func(Revision) error) error
+}
+----
+
+`memdb.go`:
+
+[source,Go]
+----
+type DB struct {
+...
+}
+...
+// Revisions implements a method of kcd.Reader.
+func (db DB) Revisions(_ context.Context, want *kcd.RevisionsFilter, f func(kcd.Revision) error) error {
+----
+
+Here we see the top level `Reader` interface again from `kcd.go`. The interface
+function `Revisions` from that same file has a `childof` edge pointing to the
+interface's node. The implementation of `Reader` that lives in `memdb.go` has a
+`satisfies` edge back to the parent interface, and its own `Revisions` function
+has an `overrides` edge pointing to the top level `Revisions` function from
+`kcd.go`. For C++ and Java, interfaces, abstract classes, and so on, have
+similar relationships in the Kythe schema.
+
+Note: For languages with explicitly-declared inheritance and interface
+implementation (such as Java), Kythe uses the `extends` edge label instead of
+`satisfies`.
+
+=== [[FeatureFunc]] Functions and Parameters
+
+Functions are represented by semantic nodes that have additional edges to
+indicate their relationship to the parameters of the function. These edges are
+named `param.0`, `param.1`, etc., where the ordinal reflects the order of
+declaration.
+
+As illustrated in this example, the parameters in turn may have their own
+declaration sites, following the conventions of the jump-to-definition example:
```

Comment:
We should explicitly note that declarations also use `defines` edges before this point.

Reply:
I'm not sure I follow why it matters here—the example that follows uses only `defines/binding`.

Reply:
Because the document talks multiple times of declarations but only implies the connection to `defines` edges.

Reply:
Ah, I see what you mean. Yes, that makes sense. PTAL.